File: 3851-find-sum-of-array-product-of-magical-sequences/find-sum-of-array-product-of-magical-sequences.java
class Solution {
    public int magicalSum(int m, int k, int[] nums) {
        int[][] comb = getComb(m, m);
        Integer[][][][] mem = new Integer[m + 1][k + 1][nums.length + 1][m + 1];
        return dp(m, k, 0, 0, nums, mem, comb);
    }

    private static final int MOD = 1_000_000_007;

    private int dp(int m, int k, int i, int carry, int[] nums,
                   Integer[][][][] mem, int[][] comb) {
        if (m < 0 || k < 0 || (m + Integer.bitCount(carry) < k))
            return 0;
        if (m == 0) {
            return (k == Integer.bitCount(carry)) ? 1 : 0;
        }
        if (i == nums.length) {
            return 0;
        }
        if (mem[m][k][i][carry] != null) {
            return mem[m][k][i][carry];
        }
        long res = 0;
        for (int count = 0; count <= m; count++) {
            long contribution = (long) comb[m][count] * modPow(nums[i], count) % MOD;
            int newCarry = carry + count;
            res = (res +
                   dp(m - count, k - (newCarry % 2), i + 1, newCarry / 2, nums, mem, comb)
                   * contribution) % MOD;
        }
        mem[m][k][i][carry] = (int) res;
        return (int) res;
    }

    private int[][] getComb(int n, int k) {
        int[][] comb = new int[n + 1][k + 1];
        for (int i = 0; i <= n; i++) {
            comb[i][0] = 1;
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= k; j++) {
                comb[i][j] = (comb[i - 1][j] + comb[i - 1][j - 1]) % MOD;
            }
        }
        return comb;
    }

    private long modPow(long x, long e) {
        long res = 1;
        x %= MOD;
        while (e > 0) {
            if ((e & 1) != 0) {
                res = (res * x) % MOD;
            }
            x = (x * x) % MOD;
            e >>= 1;
        }
        return res;
    }
}

File: 3851-find-sum-of-array-product-of-magical-sequences/find-sum-of-array-product-of-magical-sequences.js
const MOD = 1000000007;

function magicalSum(m, k, nums) {
  const n = nums.length;
  const comb = getComb(m);
  const mem = Array.from({ length: m + 1 }, () =>
    Array.from({ length: k + 1 }, () =>
      Array.from({ length: n + 1 }, () =>
        Array(m + 1).fill(undefined)
      )
    )
  );

  function dp(mr, kr, i, carry) {
    if (mr < 0 || kr < 0 || mr + bitCount(carry) < kr) return 0;
    if (mr === 0) return kr === bitCount(carry) ? 1 : 0;
    if (i === n) return 0;

    const memoVal = mem[mr][kr][i][carry];
    if (memoVal !== undefined) return memoVal;

    let res = 0n;
    for (let cnt = 0; cnt <= mr; cnt++) {
      const contribution =
        (BigInt(comb[mr][cnt]) * BigInt(modPow(nums[i], cnt))) % BigInt(MOD);
      const newCarry = carry + cnt;
      const sub = dp(mr - cnt, kr - (newCarry % 2), i + 1, Math.floor(newCarry / 2));
      res = (res + (BigInt(sub) * contribution) % BigInt(MOD)) % BigInt(MOD);
    }

    mem[mr][kr][i][carry] = Number(res);
    return Number(res);
  }

  return dp(m, k, 0, 0);
}

function getComb(n) {
  const comb = Array.from({ length: n + 1 }, () => Array(n + 1).fill(0));
  for (let i = 0; i <= n; i++) comb[i][0] = 1;
  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= i; j++) {
      comb[i][j] = (comb[i - 1][j] + comb[i - 1][j - 1]) % MOD;
    }
  }
  return comb;
}

function modPow(x, e) {
  let res = 1n;
  let base = BigInt(x % MOD);
  let exp = BigInt(e);
  while (exp > 0n) {
    if (exp & 1n) res = (res * base) % BigInt(MOD);
    base = (base * base) % BigInt(MOD);
    exp >>= 1n;
  }
  return Number(res);
}

function bitCount(x) {
  let cnt = 0;
  while (x > 0) {
    cnt += x & 1;
    x >>= 1;
  }
  return cnt;
}

console.log(magicalSum(2, 1, [1, 2, 3])); 
